File: apps/sdkwork-appstore-flutter-mobile/packages/sdkwork_appstore_flutter_mobile_category/lib/src/state/category_state.dart
import '../models/category_models.dart';

enum CategoryStatus { idle, loading, ready, error }

class CategoryState {
  const CategoryState({
    this.status = CategoryStatus.idle,
    this.lastResult,
    this.error,
  });

  final CategoryStatus status;
  final CategoryPageResult<Object?>? lastResult;
  final String? error;

  CategoryState loading() => const CategoryState(status: CategoryStatus.loading);

  CategoryState loaded(CategoryPageResult<Object?> result) =>
      CategoryState(status: CategoryStatus.ready, lastResult: result);

  CategoryState failed(String message) =>
      CategoryState(status: CategoryStatus.error, error: message);
}

File: apps/sdkwork-appstore-mini-program/packages/sdkwork-appstore-mp-category/src/state/categoryState.ts
import type { CategoryPageResult } from "../types/categoryModels.js";

export interface CategoryState {
  readonly status: "idle" | "loading" | "ready" | "error";
  readonly lastResult?: CategoryPageResult<unknown>;
  readonly error?: string;
}

export function initialCategoryState(): CategoryState {
  return { status: "idle" };
}

export function reduceCategoryState(
  state: CategoryState,
  event:
    | { type: "load" }
    | { type: "loaded"; result: CategoryPageResult<unknown> }
    | { type: "failed"; error: string },
): CategoryState {
  switch (event.type) {
    case "load":
      return { ...state, status: "loading", error: undefined };
    case "loaded":
      return { status: "ready", lastResult: event.result };
    case "failed":
      return { ...state, status: "error", error: event.error };
  }
}
